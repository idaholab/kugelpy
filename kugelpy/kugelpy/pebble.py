'''
@authors: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''

import numpy as np
import math

class Pebble(object):
    """
    Class for holding information for an individual pebble.
    
    Parameters
    ----------
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume: int
        Integer representation the volume region the pebble is in 
    channel: int
        Integer representation the channel the pebble is in     
    
    Attributes
    ----------
    pebble_type: str
        Designation if this is a graphite or fuel pebble
    inner_radius: float
        Inner radius of the pebble surrounding the graphite matrix [cm]
    material: dict
        Dictionary of the material in the pebble. Takes the form:
        {'material_name': {'ZAID1': float, 'ZAID2': float, ...}}
    geometry: dict
        Dictionary of the geometry in the pebble. 
        Each entry describes a part of the geometry, where a given radius and number of instances are used to calculate the volume
        Takes the form:
        {'geometry_name': {'radius': float, 'number_of_instances': int, 'volume': float}}
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume_num: int
        Integer representation the volume region the pebble is in 
    channel_num: int
        Integer representation the channel the pebble is in 
    mesh_location: str
        Combination of channel, and volume for where the pebble is
    xs_library: str
        Cross-section library for Serpent to use 
    scattering_library: str
        Name of the scattering library for Serpent to use
    temperature: float
        Temperature of the pebble [K]
    shuffled: Bool
        Indicator if this is the first pass or Nth pass through the core
    previous_universe: str
        Concatinated string containing the previous universes the pebble was exposed to
        If the pebble is homogenized after each pass, we default the previous universes to c0v0
    burnup: float
        Burnup accrued during pebble movement [MWd/kg]
    num_passes: int
        number of passes a pebble has gone through the core
    pebble_number: int
        integer related to the pebble number in the core
    universe: str
        string containing two integers, channel and volume
    """

    def __init__(self, x, y, z, r, radius, channel, volume, temperature=900.0, pass_limit=6, pebble_number=0):
        self._pebble_type = 'graphite'
        self._inner_radius = 2.5
        self._radius = radius
        self._material = {'matrix':   {'6000': 9.22570E-2, '5010': 2.28336E-9, '5011': 9.24876E-9},
                         'pebshell': {'6000': 6.61814E-2, '5010': 2.23273E-8,  '5011': 9.04368E-7}}
#        self._material = {'matrix':   {'6000': 9.22570E-2, },
#                         'pebshell': {'6000': 6.61814E-2, }}        
        self._x = x
        self._y = y
        self._z = z
        self._r = r
        self._volume_num = volume
        self._channel_num = channel
        self._mesh_location = f'c{channel}v{volume}'
        self._temperature = temperature
        self._shuffled = False
        self._previous_universe = ''
        self._burnup = 0.
        self._burnup_j_cm3 = 0.
#        self._burnup_group = 0
        self._num_passes = 0
        self._pass_limit = pass_limit
        self._pebble_number = pebble_number
        self._xs_dict = {1500: {'graphite': 'grph1500', 'xs_set': '15c'},
                        1200: {'graphite': 'grph1200', 'xs_set': '12c'},
                         900: {'graphite':  'grph900', 'xs_set': '09c'},
                         600: {'graphite':  'grph600', 'xs_set': '06c'},
                         300: {'graphite':  'grph300', 'xs_set': '03c'},}
        self._previous_universe = 'g0_' + self._mesh_location
        self.set_universe()
        self._xs_library, self._scattering_library = self.set_xs_set(self._temperature)

        self._geometry = {'matrix':    {'radius': self._inner_radius,    'number_of_instance': 1, 'volume': 0.0},
                         'pebshell':  {'radius': radius, 'number_of_instance': 1, 'volume': 0.0}}
        self.calculate_volumes()

    def calculate_volumes(self):
        """
        Calculate the volume for each constituent part of the pebble
        """
        previous_volume = 0.0
        for data in self._geometry.values():
            volume = math.pi * 4 / 3 * math.pow(data['radius'], 3) 
            total_region_volume = volume * data['number_of_instance']
            data['volume'] = total_region_volume - previous_volume
            previous_volume = total_region_volume 

    def set_xs_set(self, temperature):
        """
        Grab the correct cross-section set and scattering library based on the temperature
        """
        dict_array = np.array([x for x in self._xs_dict.keys()])
        temp = dict_array[dict_array<temperature].max() if temperature not in self._xs_dict.keys() else temperature
        return self._xs_dict[temp]['xs_set'], self._xs_dict[temp]['graphite']
        
    def update_position(self, x, y, z, r, channel, volume, shuffled=False):
        """
        Update the position of the pebbles after a shift or refuel.
        """
        self._shuffled = shuffled
        self._x = x
        self._y = y
        self._z = z
        self._r = r
        self._channel_num = channel
        self._volume_num = volume
        self._mesh_location = f'c{channel}v{volume}'
        self.set_universe()

    def set_previous_universe(self, prev_universe):
        """
        Update the previous universe after a shift or refuel
        """
        self._previous_universe = prev_universe

    def increase_pass(self):
        """
        When pebble is refueled, increse the number of passes it has gone through.
        """
        self._num_passes += 1
    
    def set_universe(self):
        """
        Update the current universe of the pebbble, if this is not the first pass (i.e. it has been shuffled, add the previous universe to the current pebble location
        """
        if self._shuffled:
            self._universe = self._previous_universe + '_' + self._mesh_location
        else:
            split_univ = self._previous_universe.split('_')[1:]
            univ = '' if len(split_univ) == 1 else '_'.join([x for x in split_univ[:-1]]) + '_'
            self._universe = f'g_' + univ + self._mesh_location 
            #self._universe = f'g_' + f'p{self._num_passes}_' + f'bu{self._burnup_group}_ + univ + self._mesh_location
    
    def update_pebble_temperature(self,pebble_temp,fuel_temp=None):
        """
        Update the pebble temperature and corresponding cross-section/scattering libraries.
        """
        self._temperature = pebble_temp
        self._xs_library, self._scattering_library = self.set_xs_set(pebble_temp)

class FuelPebble(Pebble):
    """
    Class for holding information for an individual pebble.
    
    Parameters
    ----------
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume: int
        Integer represention the volume region the pebble is in 
    channel: int
        Integer represention the channel the pebble is in 
    fuel_material: dict
        Dictionary containing the fuel material, will replace 'fuel' in self._material
    
    
    Attributes
    ----------
    kernels_per_pebble: int
        Number of TRISO kernels per fuel pebble
    triso_volume: float
        Volume of the fuel kernels in the pebble (required for correct burnup)
    power_days: float
        Number of megawatt days the pebbles has been in the core (days * power) [MWd]
    power_density: float
        Current power density for the pebble (this is the power density for all pebbles of the same fuel type, same pass number, and in the same core location) [MW]
    geometry: dict

    """
    
    def __init__(self, x, y, z, r, radius, channel, volume, fuel_material=None, homogenization_group=0, temperature=900.0,  fuel_temperature=900.0, pass_limit=6, pebble_number=0, kernel_data=None):
        self._homogenization_group = homogenization_group
        super().__init__(x, y, z, r, radius, channel, volume, temperature=temperature, pass_limit=pass_limit, pebble_number=pebble_number)
        self._pebble_type = 'fuel'
        self._kernel_data = kernel_data if kernel_data else {'fuel':0.02125, 'buffer':0.03125, 'inner_pyc':0.03525, 'sic':0.03875, 'outer_pyc':0.04275, 'kernels_per_pebble': 18775}
        self._kernels_per_pebble = self._kernel_data.pop('kernels_per_pebble')
        self._material = {'fuel':      {'92235': 3.70100E-3, '92238': 1.99216E-2,  '8016': 3.37331E-1, '6000': 9.26006E-3, '5010': 1.87091E-8, '5011': 7.57813E-8},
                         'buffer':    { '6000': 5.26466E-2,  '5010': 1.35512E-8,  '5011': 5.48894E-8},
                         'inner_pyc': { '6000': 9.52653E-2,  '5010': 2.45213E-8,  '5011': 9.93236E-8},
                         'sic':       {'14028': 4.43271E-2, '14029': 2.25082E-3, '14030': 1.48376E-3, '6000': 4.80616E-2,  '5010': 1.23711E-8,  '5011': 5.0109E-8},
                         'outer_pyc': { '6000': 9.52653E-2,  '5010': 2.45213E-8,  '5011': 9.93236E-8},
                         'matrix':    { '6000': 8.67416E-2,  '5010': 2.23273E-8,  '5011': 9.04368E-8},
                         'pebshell':  { '6000': 8.67416E-2,  '5010': 2.23273E-8,  '5011': 9.04368E-8}}

        #self._material = {'fuel':      {'92235': 3.70100E-3, '92238': 1.99216E-2,  '8016': 3.37331E-1, '6000': 9.26006E-3, },
        #                 'buffer':    { '6000': 5.26466E-2, },
        #                 'inner_pyc': { '6000': 9.52653E-2,  },
        #                 'sic':       {'14028': 4.43271E-2, '14029': 2.25082E-3, '14030': 1.48376E-3, '6000': 4.80616E-2, },
        #                 'outer_pyc': { '6000': 9.52653E-2, },
        #                 'matrix':    { '6000': 8.67416E-2, },
        #                 'pebshell':  { '6000': 8.67416E-2, }}                         
        self._fuel_temperature = fuel_temperature
        self._previous_universe = f'f{self._homogenization_group}_' + self._mesh_location

        self.set_universe()
        self._previous_universe = self._universe
        self._triso_volume = self._kernels_per_pebble * 0.0000402
        self._power_days = 0.
        self._power_density = 0.
        self._days_in_core = 0
        if fuel_material:
            self._material['fuel'] = fuel_material
        self._xs_fuel_library, self._fuel_scattering_library = self.set_xs_set(self._fuel_temperature)

        self._geometry = {'fuel':      {'radius': self._kernel_data['fuel'],      'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'buffer':    {'radius': self._kernel_data['buffer'],    'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'inner_pyc': {'radius': self._kernel_data['inner_pyc'], 'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'sic':       {'radius': self._kernel_data['sic'],       'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'outer_pyc': {'radius': self._kernel_data['outer_pyc'], 'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'matrix':    {'radius': self._inner_radius, 'number_of_instance': 1},
                         'pebshell':  {'radius': radius, 'number_of_instance': 1}}
        self.calculate_volumes()
        

    def update_fuel_material(self, material):
        """
        Update the fuel material
        """
        self._material['fuel'] = material

    def update_burnup(self, power, days):
        """
        Add up the total power days the pebble has been in the core.
        Divide by the kg of HM originally in the pebble (currently assumed to be 7 g or 0.007 kg)
        To obtain the burnup in J/cm^3 
        """
        self._days_in_core += days
        self._power_density = power / (1E6) # Convert from W to MW, power of an individual pebble
        self._power_days += (self._power_density * days)
        self._burnup =  self._power_days / 0.007 # MWd/kg
        self._burnup_j_cm3 = self._power_days / self._triso_volume * 1E6 * 86400# for J/cm^3 we divide the power days (Watt-day or J/s * days) by the kernel volume (cm^3) and convert to MJ from J (1E6) and from days to seconds (86,400 s/day)
        
    def set_universe(self):
        """
        Update the current universe of the pebbble, if this is not the first pass (i.e. it has been shuffled, add the previous universe to the current pebble location        
        """
        if self._shuffled:
            self._universe = self._previous_universe + '_' + self._mesh_location
        else:
            split_univ = self._previous_universe.split('_')[1:]
            univ = '' if len(split_univ) == 1 else '_'.join([x for x in split_univ[:-1]]) + '_'
            self._universe = f'f{self._homogenization_group}_' + univ + self._mesh_location   
            
    def update_pebble_temperature(self,pebble_temp,fuel_temp=None):
        """
        Update the temperatures (and corresponding cross-sections) for both the pebble and fuel materials.
        """
        self._temperature = pebble_temp
        self._fuel_temperature = fuel_temp
        self._xs_fuel_library, self._fuel_scattering_library = self.set_xs_set(fuel_temp)
        self._xs_library, self._scattering_library = self.set_xs_set(pebble_temp)
