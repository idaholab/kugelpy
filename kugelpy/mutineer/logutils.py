'''Utility module to create and write log file or string.

Created on Jan 27, 2022

@authors: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''

#===============================================================================
# Import
#===============================================================================
from os import path, makedirs
from time import strftime
from time import time

#===============================================================================
# Definitions
#===============================================================================


class LogTracker(object):
    '''Initialize log utility function.

    Parameters
    ----------
    basestr: str, optional
        Base string to print.
    fllchar: str, optional
        Filler char for the line.
    cmmchar: str, optional
        Comment char at the beginning of the line.
    NTab: int, optional
        Number of tabulation between comment and beginning of the string.
    ttlen: int, optional
        Maximum number of columns per line.
    strflag: bool, optional
        If True push into log string the just printed line.
    filflag: bool, optional
        If False, not writing to file id True it write to log_path.
    mltplflag: bool, optional
        If False, the line is cut to ttlen; the line is split into
        multiple line otherwise.
    log_path: str, optional
        Path to the log output file.

    Notes
    -----
    1: if 'time' inputed max length must be defined at least len($   15:16:15)
    2: if fllchar is more than one char it does not work

    '''

    def __init__(self, basestr:str='', fllchar:str='', cmmchar:str='#', NTab:int=0,
                 ttlen:int=100, strflag:bool=False, filflag:bool=False,
                 mltplflag:bool=False, log_path:str='log.out'):

        self.basestr = basestr
        self.fllchar = fllchar
        self.cmmchar = cmmchar
        self.NTab = NTab
        self.ttlen = ttlen
        self.strflag = strflag
        self.filflag = filflag
        self.mltplflag = mltplflag
        self.log_path = log_path
        self.log = ''
        self.timing = None

    def lprint(self, basestr=None, fllchar=None, cmmchar=None, NTab=None,
             ttlen=None, strflag=None, filflag=None, mltplflag=None, log_path=None):
        '''Print string with comment and filler.

        Parameters
        ----------
        Same as input parameters.

        '''
        # Replace default values if needed.
        if basestr or basestr == '': basestr_tmp = basestr
        else: basestr_tmp = self.basestr
        if fllchar or fllchar == '': fllchar_tmp = fllchar
        else: fllchar_tmp = self.fllchar
        if cmmchar or cmmchar == '': cmmchar_tmp = cmmchar
        else: cmmchar_tmp = self.cmmchar
        if NTab or NTab == 0: NTab_tmp = NTab
        else: NTab_tmp = self.NTab
        if ttlen: ttlen_tmp = ttlen
        else: ttlen_tmp = self.ttlen
        if strflag != None: strflag_tmp = strflag
        else: strflag_tmp = self.strflag
        if filflag != None: filflag_tmp = filflag
        else: filflag_tmp = self.filflag
        if mltplflag != None: mltplflag_tmp = mltplflag
        else: mltplflag_tmp = self.mltplflag
        if log_path != None: log_path_tmp = log_path
        else: log_path_tmp = self.log_path
        # Calculate how many char cut.
        cutchar = 0
        if (cmmchar_tmp): cutchar = cutchar + len(cmmchar_tmp) + 1  # '# '
        if (fllchar_tmp): cutchar = cutchar + 2  # ' -'
        sngltab = '  '
        tabstr = sngltab * NTab_tmp
        cutchar = cutchar + len(tabstr)
        # Calculate length and limitR length.
        strlen, maxlen = len(basestr_tmp), ttlen_tmp - cutchar
        # Prepare base string (cut or split).
        if basestr_tmp == 'time':
            basestr_tmp = self.comm_fill(strftime("%d/%m/%Y %H:%M:%S"), tabstr, \
                                        cmmchar_tmp, fllchar_tmp, ttlen_tmp)
        elif strlen >= maxlen:
            # Split string.
            if mltplflag_tmp:
                # Splitting string.
                linlist = [ basestr_tmp[0: maxlen]]
                basestr_tmp = basestr_tmp[maxlen:]
                linlist += [ basestr_tmp[i:i + maxlen - len(sngltab)] for i in
                           range(0, len(basestr_tmp), maxlen - len(sngltab)) ]
                # Add comment and filler to each line.
                basestr_tmp = ''
                # Push each line into base string.
                for i, line in enumerate(linlist):
                    basestr_tmp += ('\n' if i != 0 else '') + \
                    self.comm_fill(line, tabstr + (sngltab if i != 0 else '') , cmmchar_tmp, fllchar_tmp, \
                                  ttlen_tmp)
            # Cut string.
            else:
                # Push into base string.
                basestr_tmp = self.comm_fill(basestr_tmp[0:maxlen], tabstr,
                                            cmmchar_tmp, fllchar_tmp, ttlen_tmp)
        else:
            basestr_tmp = self.comm_fill(basestr_tmp, tabstr, cmmchar_tmp, \
                                        fllchar_tmp, ttlen_tmp)
        # Print string.
        print(basestr_tmp)
        # Push into the log string.
        if strflag_tmp:
            self.log = self.log + ('\n' if self.log else '') + basestr_tmp
        # Print to file.
        if filflag_tmp:
            if not path.isdir(path.dirname(log_path_tmp)):
                makedirs(path.dirname(log_path_tmp))
            with open(log_path_tmp, 'a') as lo:
                lo.write(basestr_tmp + '\n')

    def comm_fill(self, basestr, tabstr, cmmchar, fllchar, ttlen):
        '''Add comment and filler to a string with defined length.

        Parameters
        ----------
        basestr: str
            Base string to print.
        tabstr: str
            Total space before (multiple tabs) string.
        cmmchar: str
            Comment char at the beginning of the line.
        fllchar: str
            Filler char for the line.
        ttlen: int
            Maximum number of columns per line.

        '''
        # Add comment.
        basestr = (cmmchar if cmmchar else '') + \
        (' ' if basestr else '') + tabstr + basestr
        # Add filler.
        if fllchar:
            basestr = basestr + ' ' + \
            fllchar * (ttlen - len(basestr) - 1)
        # Return the string with comment and filler.
        return basestr

    def str_out(self):
        '''Return log string as output argument.'''
        return self.log

    def str_reset(self):
        '''Reset log string.'''
        self.log = ''

    def cnvrt_sec(self, time):
        '''Convert seconds to days hours minutes and seconds format

        Parameters
        ----------
        time:
            Time in seconds.

        Returns
        -------
        str:
            String with day, hour, minutes, seconds.

        '''
        day = time // (24 * 3600)
        time = time % (24 * 3600)
        hour = time // 3600
        time %= 3600
        minutes = time // 60
        time %= 60
        seconds = time
        return "%02d:%02d:%02d:%02d" % (day, hour, minutes, seconds)

    def start(self):
        '''Start logging message.'''
        self.lprint('Starting activities', '+', NTab=0)
        self.lprint('time', '', NTab=0)
        self.lprint('', fllchar='')
        self.timing = time()

    def stop(self):
        '''Stop logging message.'''
        self.lprint('', fllchar='')
        self.lprint('End of activities', '+', NTab=0)
        self.lprint('time', '', NTab=0)
        if self.timing:
            self.lprint('', '', NTab=0)
            self.lprint('Execution time(DD:HH:MM:SS)', '+', NTab=0)
            self.lprint(self.cnvrt_sec(time() - self.timing), '', NTab=0)
        self.lprint('', '', NTab=0, cmmchar='')
        self.timing = None

#===============================================================================
# LOG TEXT PRINTING
#===============================================================================
class logln(object):

    #----------------------------------------------------------------- Arguments
    # basestr = base string to print
    # fllchar = filler char for the line
    # cmmchar = comment char at the beginning of the line
    # NTab    = NTab number of tabulation between comment and beginning of sting
    # ttlen  = maximum number of columns per line
    # strflag = if True push into log string the just printed line
    # mltplflag = if False the line is cut otherwise multiple line plot
    #----------------------------------------------------------------------- WIP
    # 1 - if 'time' inputed max length must be defined at least len($   15:16:15)
    # 2 - if fllchar is more than one char is a mess
    #------------------------------------------ Initialize object & default data
    def __init__(self, basestr='', fllchar='', cmmchar='#', NTab=0,
                 ttlen=100, strflag=False, mltplflag=False, timing=None):
        self.basestr = basestr
        self.fllchar = fllchar
        self.cmmchar = cmmchar
        self.NTab = NTab
        self.ttlen = ttlen
        self.strflag = strflag
        self.mltplflag = mltplflag
        self.log = ''
        self.timing = timing

    #-------------------------------------- Print string with comment and filler
    def prnt(self, basestr=None, fllchar=None, cmmchar=None, NTab=None,
             ttlen=None, strflag=None, mltplflag=None, timing=None):
        #-------------------------------------- Replace default values if needed
        if basestr or basestr == '': basestr_tmp = basestr
        else: basestr_tmp = self.basestr
        if fllchar or fllchar == '': fllchar_tmp = fllchar
        else: fllchar_tmp = self.fllchar
        if cmmchar or cmmchar == '': cmmchar_tmp = cmmchar
        else: cmmchar_tmp = self.cmmchar
        if NTab or NTab == 0: NTab_tmp = NTab
        else: NTab_tmp = self.NTab
        if ttlen: ttlen_tmp = ttlen
        else: ttlen_tmp = self.ttlen
        if strflag != None: strflag_tmp = strflag
        else: strflag_tmp = self.strflag
        if mltplflag != None: mltplflag_tmp = strflag
        else: mltplflag_tmp = self.mltplflag
        if timing != None: timing_tmp = timing
        else: timing_tmp = self.timing
        #------------------------------------------- Calculate how many char cut
        cutchar = 0
        if (cmmchar_tmp): cutchar = cutchar + len(cmmchar_tmp) + 1  # '# '
        if (fllchar_tmp): cutchar = cutchar + 2  # ' -'
        tabstr = '  ' * NTab_tmp
        cutchar = cutchar + len(tabstr)
        #------------------------------------ Calculate length and limitR length
        strlen, maxlen = len(basestr_tmp), ttlen_tmp - cutchar
        #------------------------------------ Prepare base string (cut or split)
        if basestr_tmp == 'time':
            import time
            basestr_tmp = self.commfill(time.strftime("%d/%m/%Y %H:%M:%S"), tabstr, \
                                        cmmchar_tmp, fllchar_tmp, ttlen_tmp)
        else:
            if timing_tmp == 's':
                maxlen -= 6
                suffix = ' START'
            elif timing_tmp == 'e':
                maxlen -= 4
                suffix = ' END'
            else:
                suffix = ''

            if strlen >= maxlen:
                #-------------------------------------------------- Split string
                if mltplflag_tmp:
                    #------------------------------------------------ add suffix
                    basestr_tmp += suffix
                    #------------------------------------------ Splitting string
                    linlist = [ basestr_tmp[i:i + maxlen] for i in
                               range(0, strlen, maxlen) ]
                    #----------------------- Add comment and filler to each line
                    basestr_tmp = self.commfill('', tabstr, cmmchar_tmp, \
                                                fllchar_tmp, ttlen_tmp)
                    #--------------------------- Push each line into base string
                    for i, line in enumerate(linlist):
                        basestr_tmp = basestr_tmp + '\n' + \
                        self.commfill(line, tabstr, cmmchar_tmp, fllchar_tmp, \
                                      ttlen_tmp)
                #---------------------------------------------------- Cut string
                else:
                    #------------------------------------- Push into base string
                    basestr_tmp = self.commfill(basestr_tmp[0:maxlen] + suffix, tabstr,
                                                cmmchar_tmp, fllchar_tmp, ttlen_tmp)
            else:
                basestr_tmp = self.commfill(basestr_tmp + suffix, tabstr, cmmchar_tmp, \
                                            fllchar_tmp, ttlen_tmp)
        #---------------------------------------------------------- Print string
        print(basestr_tmp)
        #---------------------------------------------- Push into the log string
        if strflag_tmp:
            self.log += ('\n' if self.log else '') + basestr_tmp

    #-------------------- Add comment and filler to a string with defined length
    def commfill(self, basestr, tabstr, cmmchar, fllchar, ttlen):
        #----------------------------------------------------------- Add comment
        basestr = (cmmchar if cmmchar else '') + \
        (' ' if basestr else '') + tabstr + basestr
        #------------------------------------------------------------ Add filler
        if fllchar:
            basestr = basestr + ' ' + \
            fllchar * (ttlen - len(basestr) - 1)
        #----------------------------- Return the string with comment and filler
        return basestr

    #-------------------------------------- Return log string as output argument
    def logout(self):
        return self.log

    #---------------------------------------------------------- Reset log string
    def logreset(self):
        self.log = ''
#===============================================================================


#===============================================================================
# Main
#===============================================================================
if __name__ == '__main__':
    log = LogTracker(fllchar='', mltplflag=True, NTab=1, cmmchar='@INF@')
