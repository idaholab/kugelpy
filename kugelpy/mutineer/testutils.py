'''Utility module for testing modules.

Created on 2023-05-11 16:13:01

@authors: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''
#===============================================================================
# Import
#===============================================================================
from _ctypes import ArgumentError
from os import path, mkdir,getenv
from errno import EEXIST

#===============================================================================
# Definitions
#===============================================================================
def gen_tmp_folder(main_tests_dir):
    if not path.isdir(main_tests_dir):
        try:
            mkdir(main_tests_dir)
        except OSError as exc:
            if exc.errno != EEXIST:
                raise
            pass

def find_pbed_input(file, file_path):
    '''
    Attempts to locate the start, end, and format of Serpent pbed position input in file.

    file: (str) name of file with pbed data 
    file_path: (str) path to file
    '''

    found_start = False # keeps track of whether or not start index had been found
    start_line = None # stores fist pbed geometry input line index
    end_line = None # stores last pbed geometry input line index
    format = None # indicates format of pbed geometry input line

    input = open(path.join(file_path, file)).readlines()
    
    for i in range(len(input)):
        temp = input[i].split()
        if len(temp)==6:
            try:
                float(temp[5])
            except:
                if not found_start:
                    found_start = True
                    format = 1
                    start_line = i
        elif len(temp)==5:
            try:
                float(temp[4])
            except:
                if not found_start:
                    found_start = True
                    start_line = i
                    format = 0
        else:
            if found_start:
                end_line = i
    
    return start_line, end_line, format

def compare_pbeds(solution_file, solution_path, test_file, test_path, solution_start_line=0, solution_end_line=None, format=0):
    '''
    Reads in the position (x, y, z) and radius of pebbles for comparison with expectation.
    In some edge cases, variable rounding due to numpy version can result in pebbles being placed in different 
    universes, thus universe names are not accounted for.

    solution_file: (str) file name of pre-generated solution output used for comparison
    solution_path: (str) path to solution output file
    test_file: (str) file name of pre-generated test output used for comparison
    test_path: (str) path to test output file
    solution_start_line: (str) index of line in solution file where pebble positions begin (0 by default)
    solution_end_line: (str) index of line in solution file where pebble positions end (None by default)
    format: (int) 1 if lines are numbered, 0 otherwise
    '''

    # read in solution and test files
    solution = open(path.join(solution_path, solution_file)).readlines()[solution_start_line:solution_end_line]
    test = open(path.join(test_path, test_file)).readlines()
    
    # remove new line characters
    while '\n' in solution:
        solution.remove('\n')
    while '\n' in test:
        test.remove('\n') 

    if len(solution) != len(test):
        # if files contain different number of pebbles, write failed
        print("compare_pbeds() Error: Different number of pebbles")
        print("Solution Length: ", len(solution))
        print("Test Length: ", len(test))
        return False
    else:
        # format solution and test lists
        for i in range(len(solution)):
            temp = solution[i].split()
            solution[i] = f'{temp[0+format]} {temp[1+format]} {temp[2+format]} {temp[3+format]}'
            temp = test[i].split()
            test[i] = f'{temp[0+format]} {temp[1+format]} {temp[2+format]} {temp[3+format]}'
        
        for line in test:
            # check to make sure every pebble in test is also in solution
            if line in solution:
                # remove line from solution to reduce number of comparison operations
                solution.remove(line)
            else:
                # pebble in line not contained in solution, write failed
                print("Incorrect line: ", line)
                print("Remaining length: ", len(solution))
                return False
        
        return True

#===============================================================================
# Main
#===============================================================================
if __name__ == '__main__':
    pass

