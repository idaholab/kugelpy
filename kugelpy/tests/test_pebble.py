'''
@author: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''

import kugelpy.kugelpy.kugelpy.pebble as pebble
from kugelpy.kugelpy.mutineer.testutils import gen_tmp_folder
from os import path

main_dir = path.dirname(path.realpath(__file__))
ifiles_dir = path.join(main_dir, 'testfiles')
main_tests_dir = path.join(main_dir,  '_tmp/')
gen_tmp_folder(main_tests_dir)

def test_init(regtest):
    peb = pebble.Pebble(111,111,111,111,111,4,5,temperature=600,pass_limit=5,pebble_number=12345)
    print(vars(peb), file=regtest)

def test_update_position(regtest):
    peb = pebble.Pebble(1,1,1,1,1,1,2,temperature=600,pass_limit=5,pebble_number=12345)
    peb.update_position(111,111,111,111,4,5,shuffled=True)
    print(vars(peb), file=regtest)

def test_increase_pass(regtest):
    peb = pebble.Pebble(1,1,1,1,1,1,2,temperature=600,pass_limit=5,pebble_number=12345)
    peb.increase_pass()
    print(vars(peb), file=regtest)

def test_update_pebble_temperature(regtest):
    peb = pebble.Pebble(1,1,1,1,1,1,2,temperature=600,pass_limit=5,pebble_number=12345)
    peb.update_pebble_temperature(547,fuel_temp=900)
    print(vars(peb), file=regtest)

def test_FuelPebble_init(regtest):
    peb = pebble.FuelPebble(111,111,111,111,111,4,5,temperature=600,fuel_temperature=576, pass_limit=5,pebble_number=12345)
    print(vars(peb), file=regtest)

def test_update_fuel_material(regtest):
    peb = pebble.FuelPebble(111,111,111,111,111,4,5,temperature=600,fuel_temperature=576, pass_limit=5,pebble_number=12345)
    peb.update_fuel_material({'fractions': {'1001': 1.0}})
    print(vars(peb), file=regtest)

def test_update_burnup(regtest):
    peb = pebble.FuelPebble(111,111,111,111,111,4,5,temperature=600,fuel_temperature=576, pass_limit=5,pebble_number=12345)
    peb.update_burnup(20, 5)
    print(vars(peb), file=regtest)

def test_update_pebble_temperature(regtest):
    peb = pebble.FuelPebble(111,111,111,111,111,4,5,temperature=600,fuel_temperature=576, pass_limit=5,pebble_number=12345)
    peb.update_pebble_temperature(547,fuel_temp=900)
    print(vars(peb), file=regtest)

def test_calculate_volumes(regtest):
    peb = pebble.Pebble(111,111,111,111,3,4,5)
    peb.calculate_volumes()
    print(peb._geometry, file=regtest)
    
    fpeb = pebble.FuelPebble(111,111,111,111,3,4,5)
    fpeb.calculate_volumes()
    print(fpeb._geometry, file=regtest)    
