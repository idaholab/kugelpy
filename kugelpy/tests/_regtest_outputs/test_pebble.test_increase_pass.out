{'_pebble_type': 'graphite', '_inner_radius': 2.5, '_radius': 1, '_material': {'matrix': {'6000': 0.092257, '5010': 2.28336e-09, '5011': 9.24876e-09}, 'pebshell': {'6000': 0.0661814, '5010': 2.23273e-08, '5011': 9.04368e-07}}, '_x': 1, '_y': 1, '_z': 1, '_r': 1, '_volume_num': 2, '_channel_num': 1, '_mesh_location': 'c1v2', '_temperature': 600, '_shuffled': False, '_previous_universe': 'g0_c1v2', '_burnup': 0.0, '_burnup_j_cm3': 0.0, '_num_passes': 1, '_pass_limit': 5, '_pebble_number': 12345, '_xs_dict': {1500: {'graphite': 'grph1500', 'xs_set': '15c'}, 1200: {'graphite': 'grph1200', 'xs_set': '12c'}, 900: {'graphite': 'grph900', 'xs_set': '09c'}, 600: {'graphite': 'grph600', 'xs_set': '06c'}, 300: {'graphite': 'grph300', 'xs_set': '03c'}}, '_universe': 'g_c1v2', '_xs_library': '06c', '_scattering_library': 'grph600', '_geometry': {'matrix': {'radius': 2.5, 'number_of_instance': 1, 'volume': 65.44984694978736}, 'pebshell': {'radius': 1, 'number_of_instance': 1, 'volume': -61.261056745000964}}}
